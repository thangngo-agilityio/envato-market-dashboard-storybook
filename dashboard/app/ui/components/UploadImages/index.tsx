// Libs
import React, { memo, useState } from 'react';
import { useDropzone } from 'react-dropzone';
import {
  Box,
  Text,
  Input,
  ResponsiveValue,
  FormLabel,
  Image,
  Flex,
  CloseButton,
} from '@chakra-ui/react';

// Components
import { Loading } from '..';

// Constants
import { IMAGES } from '@/lib/constants';

// Services
import { useImageUploader } from '@/lib/hooks';

export type TUploadImageImagesProps = {
  label: string;
  isError?: boolean;
  images?: string[];
  onUploadError: (message: string) => void;
  onChange: (value: string[]) => void;
};

const UploadImagesComponent = ({
  label,
  images = [],
  isError = false,
  onChange,
  onUploadError,
}: TUploadImageImagesProps) => {
  const [previewURL, setPreviewURL] = useState<string[]>(images);
  const { onDrop } = useImageUploader({
    onChange,
    onUploadError,
    setPreviewURL,
  });

  const { getRootProps, getInputProps, isFileDialogActive } = useDropzone({
    onDrop,
  });

  const handleRemoveImage = (index: number) => {
    const updatedImages = [...previewURL];
    updatedImages.splice(index, 1);
    setPreviewURL(updatedImages);
    onChange(updatedImages);
  };

  return (
    <Flex w="100%" flexDirection="column">
      <FormLabel
        htmlFor="file"
        color="text.secondary"
        marginInlineEnd={0}
        minW="max-content"
      >
        {label}
        <Flex mt="20px" flexDirection="column" alignItems="center" gap={3}>
          {isFileDialogActive && <Loading />}
          <Flex
            alignItems="center"
            justify="center"
            gap={3}
            w={{ base: 320, md: 474 }}
            flexWrap="wrap"
          >
            {previewURL.slice(0, 3)?.map((v, i) => (
              <Box key={v} position="relative">
                <CloseButton
                  color="common.white"
                  bg="primary.600"
                  borderRadius="50%"
                  fontSize={8}
                  size="lg"
                  w={5}
                  h={5}
                  onClick={() => handleRemoveImage(i)}
                  data-testid="del-icon"
                  position="absolute"
                  right={0}
                />
                <Image
                  w={{ base: 320, md: 150 }}
                  h={{ base: 320, md: 150 }}
                  src={v || IMAGES.SIGN_UP.url}
                  alt={IMAGES.AVATAR_SIGN_UP.alt}
                  fallbackSrc={IMAGES.SIGN_UP.url}
                  borderRadius={20}
                  objectFit="contain"
                />
              </Box>
            ))}
          </Flex>
          <Flex alignItems="center" gap={3}>
            {previewURL.slice(3, 5)?.map((v, i) => (
              <Box key={v} position="relative">
                <CloseButton
                  color="common.white"
                  bg="primary.600"
                  borderRadius="50%"
                  fontSize={8}
                  size="lg"
                  w={5}
                  h={5}
                  onClick={() => handleRemoveImage(i + 3)}
                  data-testid="del-icon"
                  position="absolute"
                  right={0}
                />
                <Image
                  w={{ base: 100, md: 150 }}
                  h={{ base: 100, md: 150 }}
                  src={v || IMAGES.SIGN_UP.url}
                  alt={IMAGES.AVATAR_SIGN_UP.alt}
                  fallbackSrc={IMAGES.SIGN_UP.url}
                  borderRadius={20}
                  objectFit="contain"
                />
              </Box>
            ))}
          </Flex>
        </Flex>
      </FormLabel>
      <Box
        {...getRootProps()}
        border="1px"
        borderColor={isError ? 'primary.1000' : 'primary.600'}
        borderRadius="md"
        p={4}
        textAlign="center"
        cursor="pointer"
        _hover={{ borderColor: 'green.500' }}
        bg="background.body.primary"
      >
        <Input
          {...getInputProps()}
          size={undefined as ResponsiveValue<string> | undefined}
          data-testid="field-image"
        />

        <Text>Drag drop some files here, or click to select files</Text>
      </Box>
    </Flex>
  );
};

const UploadImages = memo(UploadImagesComponent);

export default UploadImages;
