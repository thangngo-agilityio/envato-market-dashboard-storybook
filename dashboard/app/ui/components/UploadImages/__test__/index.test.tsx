import UploadImages from '..';

describe('Upload Product', () => {
  it('Match snapshot with default props', () => {
    const { container } = render(
      <UploadImages
        label="Upload Images"
        onUploadError={() => {}}
        onChange={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders correctly with label and input field', () => {
    const { getByText, getByTestId } = render(
      <UploadImages
        label="Upload Images"
        onUploadError={() => {}}
        onChange={() => {}}
      />,
    );

    expect(getByText('Upload Images')).toBeInTheDocument();
    expect(getByTestId('field-image')).toBeInTheDocument();
  });
});
