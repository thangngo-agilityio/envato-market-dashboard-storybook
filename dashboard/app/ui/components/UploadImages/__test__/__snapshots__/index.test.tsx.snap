// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Product Match snapshot with default props 1`] = `
<div>
  <div
    class="css-hk64jw"
  >
    <label
      class="chakra-form__label css-1rh809e"
      for="file"
    >
      Upload Images
      <div
        class="css-unhozd"
      >
        <div
          class="css-vqdf2v"
        />
        <div
          class="css-193m7l6"
        />
      </div>
    </label>
    <div
      class="css-oeoy58"
      role="presentation"
      tabindex="0"
    >
      <input
        class="chakra-input css-fmf4uc"
        data-testid="field-image"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
      <p
        class="chakra-text css-16h4vrm"
      >
        Drag drop some files here, or click to select files
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
